File: src/main/java/org/commonground/formbuilder/model/form/condition/ConditionParser.java
package org.commonground.formbuilder.model.form.condition;

import java.util.ArrayList;
import java.util.List;

import org.commonground.formbuilder.model.FormWrapper;
import org.commonground.formbuilder.model.form.Field;
import org.commonground.formbuilder.model.form.FieldType;
import org.commonground.formbuilder.model.form.Form;
import org.commonground.formbuilder.services.FormServiceLocal;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;

/**
 * Ik gebruik een vereenvoudigde JSONPath-syntaxis om logica te definiëren binnen formulieren.
 * $                        - De root identifier; verwijst naar het begin van het JSON-document.	$
 * . of []                  - Child segment; toegang tot een objecteigenschap of array-index.	$.store of $[0]
 * ..                       - Recursive descent; zoekt naar matches op alle diepteniveaus.	$..author
 * *                        - Wildcard; matcht alle elementen of velden.	$.books[*]
 * [start:end:step]         - Array slice; selecteert een reeks elementen uit een array.	$.books[0:5:2]
 * 
 * Voorbeelden van JSONPath-expressies:
 * $.winkel.locatie	        - Eenvoudig pad: Selecteert een specifieke waarde via dot-notatie.	"Utrecht"
 * $.winkel.boeken[*].titel	- Wildcard: Haalt alle titels op uit de boeken-array.	["Java Expert", "JSON Mastery", "RFC Gids"]
 * $..prijs	                - Deep scan: Zoekt alle velden met de naam "prijs", ongeacht de diepte.	[35.50, 20.00, 15.00]
 * $.winkel.boeken[0,2]	    - Multiple selectors: Selecteert het eerste en derde element (index 0 en 2).	De 1e en 3e boek-objecten.
 * $.winkel.boeken[-1:]	    - Slice (negatief): Selecteert het laatste element uit de array.	Het object voor "RFC Gids".
 */
public class ConditionParser {
    public static boolean checkLogic(Form form, Condition condition) {
        if( condition.getConditionType() == ConditionType.COMPOSITE) {
            if( condition.getLogicalOperator() == null) {
                throw new IllegalArgumentException("Composite conditions must have a logical operator.");
            }
            if( condition.getConditions() == null || condition.getConditions().isEmpty()) {
                throw new IllegalArgumentException("Composite conditions must contain at least one sub-condition.");
            }
            for(Condition subCondition : condition.getConditions()) {
                if (condition.getLogicalOperator() == LogicalOperator.AND) {
                    if (!checkLogic(form, subCondition)) {
                        return false;
                    }
                } else if (condition.getLogicalOperator() == LogicalOperator.OR) {
                    if (checkLogic(form, subCondition)) {
                        return true;
                    }
                }
                throw new IllegalArgumentException("Composite conditions must have a logical operator.");
            }
        } else {
            if( condition.getVar1() == null || condition.getVar1().isEmpty() || condition.getOperator() == null) {
                throw new IllegalArgumentException("Simple conditions must have a field and an operator.");
            }
            return evaluateCondition(form, condition);
        }
        return true;
    }

    public static boolean evaluateCondition(Form form, Condition condition) {
        String field = condition.getVar1();
        if (field.startsWith("var.")) {

            field = field.substring(4);
            String[] fields = field.split("\\.");
            if (fields.length < 2) {
                throw new IllegalArgumentException("Invalid variable path: " + condition.getVar1());
            }
            if (fields.length > 5) {
                throw new IllegalArgumentException("Variable path too deep: " + condition.getVar1());
            }
            return true;
            // return evaluatePath(form, fields, 0);
        }

        return checkValue(field, condition.getOperator(), condition.getVar2());
    }



    public static void getBase(String name, Field base, char[] path, int index) {
        if (!base.getName().equals(name)) {
            throw new IllegalArgumentException("Element " + name + " doesn't exist");
        }

    }

    public static boolean checkValue(String fieldValue, Operator operator, String conditionValue) {
        switch (operator) {
            case EQ:
                return fieldValue.equals(conditionValue);
            case NEQ:
                return !fieldValue.equals(conditionValue);
            case GT:
                return compareDouble(fieldValue, conditionValue) > 0;
            case LT:
                return compareDouble(fieldValue, conditionValue) < 0;
            case GTE:
                return compareDouble(fieldValue, conditionValue) >= 0;
            case LTE:
                return compareDouble(fieldValue, conditionValue) <= 0;
            default:
                throw new IllegalArgumentException("Unsupported operator: " + operator);
        }
    }

    public static int compareDouble(String fieldValue, String conditionValue) {
        try {
            Double fieldDouble = Double.parseDouble(fieldValue);
            Double conditionDouble = Double.parseDouble(conditionValue);
            return fieldDouble.compareTo(conditionDouble);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Values must be numeric for comparison operators.", e);
        }
    }
    
}

File: src/main/resources/static/includes/scripts/shared/condition-components/condition-parser.js
import {  ConditionType, Operator, LogicalOperator } from './types/condition-types.js';
import { JsonPathFinder } from './json-path-finder.js';
import { FormService } from '../../form-viewer/services/form-service.js';

export class ConditionParser {
    condition;
    func;

    constructor(condition, func) {
        this.condition = condition;
        this.func = func;
        
        FormService.getInstance().addEventListener((form) => {
            this.#bindFieldToConditions(this.condition, form);
        });
    }


    /**
     * This function is also called from outside
     */
    eval() {
        if (this.func && this.condition && !Object.values(this.condition).every(waarde => !waarde)) {
            
            this.func(this.checkLogic(this.condition));
        }
    }

    isField(input) {
        return input && input.startsWith("$.");
    }

    getConditionType(condition) {
        return !condition.conditions || condition.conditions.length == 0 ? ConditionType.SIMPLE : ConditionType.COMPOSITE;
    }

    #bindFieldToConditions(condition, form) {
        if (condition && form) {
            if (this.getConditionType(condition) == ConditionType.COMPOSITE) {
                for (const i in condition.conditions) {
                    this.#bindFieldToConditions(condition.conditions[i], form);
                }
            } else {
                this.#bindSimpelConditions(condition, form);
            }
        }
    }
    
    #bindSimpelConditions(condition, form) {
        if (this.isField(condition.var1)) {
            condition.var1Fields = JsonPathFinder.evalTokenized(condition.var1, form);
            this.#setFieldListeners(condition.var1Fields);
        }
        if (this.isField(condition.var2)) {
            condition.var2Fields = JsonPathFinder.evalTokenized(condition.var2, form);
            this.#setFieldListeners(condition.var2Fields);
        }
    }

    #setFieldListeners(fields) {
        for (const field of fields) {
            field.addValueChangedListener(() => {
                this.eval();
            });
        }
    }

    checkLogic(condition) {
        if (this.getConditionType(condition) == ConditionType.COMPOSITE) {
            
            if( condition.logicalOperator == null) {
                throw new Error("Composite conditions must have a logical operator.");
            }
            if( !condition.conditions || condition.conditions.length === 0) {
                return true;
            }

            if (condition.logicalOperator == LogicalOperator.AND) {
                for(const subCondition of condition.conditions) {
                    if (!this.checkLogic(subCondition)) {
                        return false;
                    }
                }
                return true;
            }
            if (condition.logicalOperator == LogicalOperator.OR) {
                for(const subCondition of condition.conditions) {
                    if (this.checkLogic(subCondition)) {
                        return true;
                    }
                }
                return false;
            }
            
        } else {
            
            if(!condition.operator) {
                throw new Error('Simple conditions must have an operator.');
            }
            
            return this.#evaluateCondition(condition);
        }
    }

    #evaluateCondition(condition) {
        let var1List = condition.var1Fields ? this.#getFieldValues(condition.var1Fields) : [condition.var1];
        let var2List = condition.var2Fields ? this.#getFieldValues(condition.var2Fields) : [condition.var2];
        for (const var1 of var1List) {
            for (const var2 of var2List) {
                if (!this.#evaluateConditionType(var1, condition.operator, var2)) {
                    return false;
                }
            }
            return true;
        }
    }


    #evaluateConditionType(var1, operator, var2) {
        const isArray1 = Array.isArray(var1);
        const isArray2 = Array.isArray(var2);

        if (isArray1 && isArray2) {
            if (var1.length !== var2.length) return false;
            return var1.every((val, index) => this.#checkValue(val.value, operator, var2[index].value));
        }

        // 2. var1 is een string, var2 is een lijst: Check of var1 in var2 voorkomt
        if (!isArray1 && isArray2) {
            return var2.some(item => this.#checkValue(var1, operator, item.value));
        }

        // 3. var1 is een lijst, var2 is een string: Check of var2 in var1 voorkomt
        if (isArray1 && !isArray2) {
            return var1.some(item => this.#checkValue(item.value, operator, var2));
        }

        return this.#checkValue(var1, operator, var2);
    }
        



    #getFieldValues(fields) {
        const output = [];

        for (const field of fields) {
            if (field.hasOptions()) {
                output.push(field.getOptions())
            } else {
                output.push(field.getValue())
            }
        }
        return output;
    }

    #checkValue(var1, operator, var2) {
        switch (operator) {
            case Operator.EQI:
                return (!var1 ? var1 : var1.toLowerCase()) == (!var2 ? var2 : var2.toLowerCase());
            case Operator.NEQI:
                return (!var1 ? var1 : var1.toLowerCase()) != (!var2 ? var2 : var2.toLowerCase());
            case Operator.EQ:
                return var1 == var2;
            case Operator.NEQ:
                return var1 != var2;
            case Operator.GT:
                return this.#getNumber(var1) > this.#getNumber(var2);
            case Operator.LT:
                return this.#getNumber(var1) < this.#getNumber(var2);
            case Operator.GTE:
                return this.#getNumber(var1) >= this.#getNumber(var2);
            case Operator.LTE:
                return this.#getNumber(var1) <= this.#getNumber(var2);
            default:
                throw new Error("Unsupported operator: " + operator);
        }
    }

    #getNumber(var1) {
        if (isNaN(var1)) {
            throw new Error('Values must be numeric for comparison operators.');
        }
        return new Number(var1);
    }
    
}
